### PR TITLE
Activation Protocol and Layer struct

### PR DESCRIPTION
**Pull request by perlfly**
Starts encapsulating Layer logic in a dedicated struct.
This could lead to an easier way for scaling the number of layers.
Uses some explicit pointers logic in order to reduce array copying.
This logic should remain limited to private logic.

### CODE REVIEW

**Comment by collinhundley** [comment]
I'll go ahead and make this merge.  I still need to go through the code in more detail, but everything seems to work okay so it should be safe for now (normally I'd do more testing before merging, but my time is limited this week so I'll let everybody else look it over too).  I'd just like to make one request, based on what I've seen so far:  @perlfly will you please include detailed documentation in the code?  This includes official 'documentation' for all functions and properties, as well as comments explaining each step in the calculations.  This will help all of us to better understand the logic and debug any issues that might come up.  It looks like some of the methods are already documented, so if you could keep expanding that, that would be great.  Thanks!

I like this direction for managing multiple layers - it should allow us to add support for many hidden layers without too much difficulty.

**Comment by perlfly** [comment]
Hi @collinhundley my FFNN_Prop was intended as a proposal that needs further work. Since it changes a lot the FFNN design. Appreciate you like the directions. Now it makes sense going along with it.
I'd recommend a dedicated branch, say FFNN_Dev or sorting like that, where contributors can share changes to the FFNN before it is completed. This without _braking_ into the master branch.
Once FFNN_Dev is more tested an documented, you could marge it into master.

**Comment by KacheFlowe** [comment]
I agree with @perlfly. If we're working on major features there should be different branches.

**Comment by collinhundley** [comment]
Now that Christmas is over, I've had a little time to sit down and look at the changes. @perlfly is right about needing another branch, so I created a new branch called FFNN_Deep_Dev. This is where we'll work on the feed-forward neural network until it fully supports multiple hidden layers, at which point I'll merge it back up to master.

I'm happy with the way this neural network is coming along - it works very well compared to most of the other neural nets I've used!  Adding support for multiple hidden layers will be a nice improvement, as well as finishing Softmax and improving the learning algorithm.  Eventually I'd like to improve the `train()` method to be a little more intuitive, and allow it to adjust `learningRate` and `momentumFactor` on the fly.  Adjusting those values based on the network's learning progress can really help it converge on a better solution, in my experience.

Due to the [vanishing gradient issue](http://neuralnetworksanddeeplearning.com/chap5.html), it's not likely that anybody will use FFNN for extremely deep learning.  I'd suspect that 3 hidden layers is probably the max before it starts to lose effectiveness. So although we'll continue our efforts to support multiple hidden layers (because they can be useful), soon I would like to start building a Convolutional Neural Network, with support for deeper learning.  That way, we can keep FFNN fairly lightweight and ideal for fast, simple applications, and gear CNN toward larger, more difficult tasks (computer vision, voice recognition, etc).  Eventually a Recurrent Neural Network will also find its way into Swift AI, but I don't want to worry about that quite yet.  Thoughts?
